ignite/metrics/ssim.py

```diff
@@ -156,8 +158,25 @@ def update(self, output: Sequence[torch.Tensor]) -> None:
             )
 
         channel = y_pred.size(1)
-        if len(self._kernel.shape) < 4:
-            self._kernel = self._kernel.expand(channel, 1, -1, -1).to(device=y_pred.device)
+        if (
+            not self._kernel.numel()  # when the tensor is still empty from the __init__
+            or self._kernel.shape[0] != channel
+        ):
+            self._kernel = self._kernel_2d.expand(channel, 1, -1, -1)
+
+        if y_pred.device != self._kernel.device:
+            if self._kernel.device == torch.device("cpu"):
+                self._kernel = self._kernel.to(device=y_pred.device)
+
+            if y_pred.device == torch.device("cpu"):
+                warnings.warn(
+                    "The metric device or one of the previous update tensor was set on another device than this "
+                    "update tensor, which is on CPU. To avoid having a performance hit, ensure that your metric "
+                    "device and all of your update tensors are on the same device.",
+                    RuntimeWarning,
+                )
```

Comment:
I think we can do here with `elif` and let's improve the warning message
```suggestion
            elif y_pred.device == torch.device("cpu"):
                warnings.warn(
                    f"y_pred tensor is on cpu device but previous computation was on another device: {self._kernel.device}. "
                    "To avoid having a performance hit, please ensure that all y and y_pred tensors "
                    "are on the same device.",
                )
```
Let's use UserWarning type instead of RuntimeWarning.